### ISSUE TITLE
Figure out how to test on Fastmail again

### ISSUE THREAD

**Issue by WhyNotHugo** [pimutils/vdirsyncer#824]
Fastmail has not-completely standard support for VTODO (they used to be standard, so that's a sad turn of events).

I've somehow got it working on my desktop, but I don't know how I hacked it together. In order to get master back to green, I've just turned that feature off until we have time to research and document it.

**Comment by dilyanpalauzov** [comment]
What is not standard in the VTODO for Fastmail (=cyrus imap)?

**Comment by stweller** [comment]
Can I help to test something? I do use vdirsyncer + todoman for todos and I do have a fastmail account.

**Comment by WhyNotHugo** [comment]
Fastmail changed some stuff during the last year and calendaring is a bit non-standard.

We used to be able to have `VEVENT`s and `VTODO`s in the same collection, but they changed that, and now collections that have one will fail to work with the other (they also split all existing collections into two).

Tests need to be reworked a bit to take this into consideration, since a lot of tests put both things into a single collection (also, there's tests to confirm if we're recognising the type right, and these can't be run here really either).

@stweller I do too! ㅅ_ㅅ 
Thanks, but we already have a test account that Fastmail sponsors for us to test integration with them/cyrus.

Final note: We're still testing `VEVENT`s on Fastmail, it's just `VTODO`s that are skipped entirely.

**Comment by robmueller** [comment]
FYI if you create a new calendar collection and set `supported-calendar-component-set` (https://tools.ietf.org/html/rfc4791#section-5.2.3) appropriately, you should be able to create a collection that holds VEVENT's and VTODO's.

**Comment by WhyNotHugo** [comment]
Setting the property is a bit tricky, since `vdirsyncer` has no way of knowing what type of entries a user will place in a collection.

A good way to approach this would be for us to set both types, as in the example given in the RFC:

```xml
<C:supported-calendar-component-set
    xmlns:C="urn:ietf:params:xml:ns:caldav">
  <C:comp name="VEVENT"/>
  <C:comp name="VTODO"/>
</C:supported-calendar-component-set>
```

Alternatively, we could set the property based on items in the collection (and update this pre-push if the item types changes). `vdirsyncer` would have to internally track the collection type for this to happen.

**Comment by WhyNotHugo** [comment]
Fastmail seems to return 403 when trying to alter this property:

Request:

```xml
<?xml version="1.0" encoding="utf-8" ?>

<propertyupdate xmlns="DAV:">
	<set>
		<prop>
      <C:supported-calendar-component-set
			 xmlns:C="urn:ietf:params:xml:ns:caldav">
				<C:comp name="VEVENT"/>
				<C:comp name="VTODO"/>
			</C:supported-calendar-component-set>
    </prop>
	</set>
</propertyupdate>
```

Response:

```xml
<?xml version="1.0" encoding="utf-8"?>
<multistatus
  xmlns="DAV:"
  xmlns:C="urn:ietf:params:xml:ns:caldav">
  <response>
    <href>/dav/calendars/user/whynothugo@fastmail.com/502708f4-4b94-4963-b042-f7340c829bdf/</href>
    <propstat>
      <prop>
        <C:supported-calendar-component-set/>
      </prop>
      <status>HTTP/1.1 403 Forbidden</status>
      <error>
        <cannot-modify-protected-property/>
      </error>
    </propstat>
  </response>
</multistatus>
```

I read the whole RFC and can't see any indication of why that might be. @robmueller any hints?

**Comment by dilyanpalauzov** [comment]
Fastmail and Cyrus Imap is almost the same.  You can log with a browser at https://mail.aegee.org/dav/calendars/user/aaa/ with username aaa and password abc .  Then you can change the supported components.  The browser sends in the background PROPPATCH.  As far as I can see it sends `<C:supported-calendar-component-set xmlns:C="urn:ietf:params:xml:ns:caldav" force="yes">` so with a `force` attribute.  In any case, you can inspect how the browser changes the components, by sending PROPPATCH and then implement this logic in vdirsyncer.

The `force` attribute is unnecessary, when during the collection creation the right parameters are sent.  The above web-client does send the  MKCOL requests to the server, based on the selected components. Just give it a try and inspect the web traffic.

**Comment by WhyNotHugo** [comment]
Ah, looks like `force=true` does it, nice!

Thanks for that tip (I can also finally use `VEVENT` and `VTODO` in Fastmail again!).

That UI is pretty handy for inspecting and testing. Is that part of cyrus?

**Comment by dilyanpalauzov** [comment]
The UI is partially part of Cyrus IMAP, partially I have added to it additional features.

**Comment by robmueller** [comment]
Just as an FYI, at Fastmail we split VTODO vs VEVENT collections because of problems we were seeing with users using iOS.

Basically (as noted) we used to allow VEVENTs and VTODOs in each collection, but in iOS VEVENT items appear in the Calendar app but VTODO items appear in the Reminders app. However both apps display the list of available collections that support each of their particular item type. That's fine, until someone sees in the Reminders app some collection that can hold VTODO items, but currently is only holding VEVENT items. In Reminders, it appears as an empty collection (because it has 0 VTODO items), so they decide to delete it.

Even though iOS provides a double warning with a text that mentions that your calendar and its events are going to be deleted, we know that people often just don't read or fully understand these warnings, and they end up deleting a calendar with lots of important events on it. Ooops! Worse, they might not notice until days/weeks later that it's gone. Then they contact support about it. If it's too late, we might not even have a backup any more, but even if we do, it's a considerably support headache just dealing with this situation and explaining that it was a mistake they made.

Because of this, we decided to separate out VEVENT and VTODO items into separate collections by default, which actually more closely matches how users see these things, as actual separate things.

As noted though, we did that just by changing the `supported-calendar-component-set`, so it's possible to change this back if you want and/or create a new collection with whatever supported types you want.

**Comment by stweller** [comment]
It's good to see that one of the co-founders of Fastmail takes the time to help out here a bit even if it's for a relatively small project. Thanks a lot, @robmueller ! Proves that it is the right choice to be a fastmail customer :)

**Comment by WhyNotHugo** [comment]
Thanks for the very clear explanation @robmueller, it's most welcome to finally have all the pieces to the puzzle!

I always value honest feedback a lot, so here's mine on the subject:

**As a Fastmail user**

When this change happened, it became impossible to create, mange or delete new VTODO collections via the Fastmail UI. They stopped showing up and can no longer be interacted with. I expected this to just be short term, but it's stayed that way for years now.

iOS's tools don't not allow changing `supported-calendar-component-set`, so it become impossible to somehow make a collection contain `VTODO`s again. Once a collection has a specific type, that's permanent from iOS's perspective.

It seems that Fastmail+iOS alone are not enough to work around this, and external tools are required.

**As a vdirsyncer developer**:

This was not clearly documented, so it's been a long time until we fully figured out how to work around this. I do appreciate the details given now. Granted `supported-calendar-component` is part of the standard, but I don't think we've seen this particular quirk on any other DAV server, so nobody though to check on that.

Ironically, I wrote [todoman](https://github.com/pimutils/todoman/#todoman), and currently maintain vdirsyncer, but could not actually leverage both tools together to push all my collections to Fastmail.

---

On the topic at hand, I believe that the sane thing for us to do to add a new metadata to for collections: `item-types`. This can contain one or more of the following `VTODO`, `VEVENT`, `VJOURNAL`. It'll behave the same way others do (`displayname`, `colour`, etc), though due to their format, they'll likely have to be special-cased in our sync logic.

`vdirsyncer` will reflect this on the remote collection, but won't change the type of a collection without a user having requested it explicitly. Users must edit this file manually if they expect any change on the remote side, and this is precisely to avoid confusing users with the scenario that Rob just described.

**Comment by dilyanpalauzov** [comment]
`CALDAV:supported-calendar-component-set` is a protected property.  A discussion on calsify@ietf.org about lifting this limitation  lead nowhere: https://mailarchive.ietf.org/arch/msg/calsify/8gGXRpU6Pg9BkaGYHzDMhKkpvPc .

**Comment by WhyNotHugo** [comment]
Already fixed via #909
